### PR TITLE
Allow uart to go ~100Hz versus 10Hz

### PR DESCRIPTION
**Pull request by dherrada**
So I tried a bunch of different ways involving checking how many bytes were in waiting, but in the end, just making the time.sleep smaller and adding a try/except loop was the fastest and most stable way to do it.

### CODE REVIEW

**Comment by tannewt** [comment]
Why have the sleep at all? Why not just wait for two bytes?

**Comment by dherrada** [comment]
@tannewt 
So I tried doing that with:
```
while self._uart.in_waiting < 2:
    pass
```
And it didn't work. It kept getting caught for some reason.

**Comment by tannewt** [comment]
You may need to add a timeout so that if it doesn't work you can escape the loop. Basically track how long has passed since you started and exit the loop if it is too long.

**Comment by dherrada** [comment]
Ah, ok, that makes sense. I'll do that.

**Comment by dherrada** [comment]
It is slightly janky, and if you've got a cleaner way to do it, let me know because I definitely think this could be a lot cleaner.

File: adafruit_bno055.py [hunk 1]
```diff
@@ -401,12 +403,27 @@ def _write_register(self, register, data):  # pylint: disable=arguments-differ
 
     def _read_register(self, register, length=1):  # pylint: disable=arguments-differ
         self._uart.write(bytes([0xAA, 0x01, register, length]))
-        time.sleep(0.1)
+        now = time.time()
+        while self._uart.in_waiting < length + 2 and time.time() - now < 0.25:
+            pass
         resp = self._uart.read(self._uart.in_waiting)
+
+        if resp[0] != 0xBB:  # Recursion
```
Comment (tannewt) [inline adafruit_bno055.py:411]: This is not recursion because recursion would call _read_register. Why did you add this? It wasn't needed before.

File: adafruit_bno055.py [hunk 2]
```diff
@@ -392,7 +392,9 @@ def _write_register(self, register, data):  # pylint: disable=arguments-differ
         if not isinstance(data, bytes):
             data = bytes([data])
         self._uart.write(bytes([0xAA, 0x00, register, len(data)]) + data)
-        time.sleep(0.1)
+        now = time.time()
```
Comment (tannewt) [inline adafruit_bno055.py:395]: Use `time.monotonic` instead of `time.time` because it'll work on all CP boards that way. `time.time` needs to know the calendar time while monotonic is relative to start up.

File: adafruit_bno055.py (hunk 1 above)
Comment (dherrada) [reply adafruit_bno055.py:411]: My initial plan was to actually do that, but I couldn't think of a good way to add in an exit condition so it didn't just keep going until it reached the recursion limit, and I guess I forgot to remove the comment.

It's needed because the, when running in UART mode, even in the version that's currently on master, it raises a UART read error every 3 or 4 reads when you're trying to use the simpletest. I think that before it may have mostly been tested by getting a single value, such as acceleration, at a time. That being said, even though I know the majority of people aren't using this with UART, I am quite surprised that I haven't seen anyone else encountering that issue, so it may be worthwhile for some other people with the sensor to do some testing to ensure that issue doesn't only affect me.

File: adafruit_bno055.py (hunk 2 above)
Comment (dherrada) [reply adafruit_bno055.py:395]: Will do

File: adafruit_bno055.py (hunk 1 above)
Comment (tannewt) [reply adafruit_bno055.py:411]: Do the errors go away if you increase the timeout? If you really need it, I think you could just have an outer loop and count how many tries you want to do before raising an error. No need to copy things a second time.

**Comment by dherrada** [comment]
Ok, I think I'll do that.

They don't. It seems it just gets stuck and the only way to get it unstuck is to re-request the data from the sensor.

**Comment by dherrada** [comment]
@tannewt Yeah, this works way better now. I felt confident enough in it to lower the timeout time to 0.1 seconds from 0.25 and I let it run for a while and it hasn't given me any issues.

File: adafruit_bno055.py [hunk 3]
```diff
@@ -392,17 +392,26 @@ def _write_register(self, register, data):  # pylint: disable=arguments-differ
...
-        resp = self._uart.read(self._uart.in_waiting)
+        i = 0
+        while i < 3:
+            self._uart.write(bytes([0xAA, 0x01, register, length]))
+            now = time.monotonic()
+            while self._uart.in_waiting < length + 2 and time.monotonic() - now < 0.1:
+                pass
+            resp = self._uart.read(self._uart.in_waiting)
+            if len(resp) >= 2 and resp[0] == 0xBB:
+                i = 3
```
Comment (tannewt) [inline adafruit_bno055.py:413]: Don't mutate `i`, just `break` to exit the loop.

File: examples/bno055_simpletest.py [hunk 4]
```diff
@@ -12,7 +12,6 @@
 # sensor = adafruit_bno055.BNO055_UART(uart)
 
 while True:
-    print("Temperature: {} degrees C".format(sensor.temperature))
```
Comment (tannewt) [inline examples/bno055_simpletest.py:15]: Why remove this?

Comment (dherrada) [reply examples/bno055_simpletest.py:15]: Oops, didn't mean to do that.
